Source/Dafny/Compiler-cpp.cs

```diff
@@ -0,0 +1,2338 @@
+      }
+    }
+
+    protected override BlockTargetWriter EmitTailCallStructure(MemberDecl member, BlockTargetWriter wr) {
+      wr.WriteLine("TAIL_CALL_START:");
+      return wr;
+    }
+
+    protected override void EmitJumpToTailCallStart(TargetWriter wr) {
+      wr.WriteLine("goto TAIL_CALL_START;");
+    }
+
+    protected void Warn(string msg, Bpl.IToken tok) {
+      Console.Error.WriteLine("WARNING: {3} ({0}:{1}:{2})", tok.filename, tok.line, tok.col, msg);
+    }
+
+    // Because we use reference counting (via shared_ptr), the TypeName of a class differs
+    // depending on whether we are declaring a variable or talking about the class itself.
+    // Use class_name = true if you want the actual name of the class, not the type used when declaring variables/arguments/etc.
+    protected string TypeName(Type type, TextWriter wr, Bpl.IToken tok, MemberDecl/*?*/ member = null, bool class_name=false) {
+      Contract.Ensures(Contract.Result<string>() != null);
+      Contract.Assume(type != null);  // precondition; this ought to be declared as a Requires in the superclass
+
+      var xType = type.NormalizeExpand();
+      if (xType is TypeProxy) {
+        // unresolved proxy; just treat as ref, since no particular type information is apparently needed for this type
+        return "object";
+      }
+
+      if (xType is BoolType) {
+        return "bool";
+      } else if (xType is CharType) {
+        return "char";
+      } else if (xType is IntType || xType is BigOrdinalType) {
+        Warn("BigInteger used", tok);
+        return "BigNumber";
+      } else if (xType is RealType) {
+        Warn("BigRational used", tok);
+        return "Dafny.BigRational";
+      } else if (xType is BitvectorType) {
+        var t = (BitvectorType)xType;
+        return t.NativeType != null ? GetNativeTypeName(t.NativeType) : "BigNumber";
+      } else if (xType.AsNewtype != null) {
+        NativeType nativeType = xType.AsNewtype.NativeType;
+        if (nativeType != null) {
+          return GetNativeTypeName(nativeType);
+        }
+        return TypeName(xType.AsNewtype.BaseType, wr, tok);
+      } else if (xType.IsObjectQ) {
+        return "object";
+      } else if (xType.IsArrayType) {
+        ArrayClassDecl at = xType.AsArrayType;
+        Contract.Assert(at != null);  // follows from type.IsArrayType
+        Type elType = UserDefinedType.ArrayElementType(xType);
+        if (at.Dims == 1) {
+          return "DafnyArray<" + TypeName(elType, wr, tok, null, false) + ">";
+        } else {
+          throw NotSupported("Multi-dimensional arrays");
+        }
+      } else if (xType is UserDefinedType) {
+        var udt = (UserDefinedType)xType;
+        var s = FullTypeName(udt, member);
+        var cl = udt.ResolvedClass;
+        bool isHandle = true;
+        if (cl != null && Attributes.ContainsBool(cl.Attributes, "handle", ref isHandle) && isHandle) {
+          return "ulong";
+        } else if (DafnyOptions.O.IronDafny &&
```

Comment:
This question is out of place here, but perhaps you know the answer: Is `DafnyOption.O.IronDafny` used by anyone anymore? If it's not needed for C++, you can remove the case here (and the same for the other back ends).

Reply:
Jay, Jon, and I are no longer using it.  I'm waiting to hear back from Manos.  If he's not, then I think it would be safe to remove at this point.